lib/TaskProcessing/SummaryProvider.php

```diff
@@ -101,34 +101,108 @@ public function process(?string $userId, array $input, callable $reportProgress)
+				throw new RuntimeException('No result in OpenAI/LocalAI response.');
+			}
+
+			// Take only one completion for each chunk and combine them into a single summary (which may be used as the next prompt)
+			$completionStrings = array_map(fn (array $val): string => end($val), $completions);
+			$summary = implode(' ', $completionStrings);
+
+			$prompts = self::chunkSplitPrompt($summary);
+			$newNumChunks = count($prompts);
+		} while ($oldNumChunks > $newNumChunks);
+
+		$endTime = time();
+		$this->openAiAPIService->updateExpTextProcessingTime($endTime - $startTime);
+		return ['output' => $summary];
+	}
+
+	private function chunkSplitPrompt(string $prompt): array {
+		$chunkSize = $this->openAiSettingsService->getChunkSize();
+
+		// https://platform.openai.com/tokenizer
+		// Rough approximation, 1 token is approximately 4 characters/bytes for OpenAI models
+		// It's better to have a lower estimate on the max number of tokens, so consider 3 bytes per token instead of 4 (to account for some multibyte characters)
+		$maxChars = $chunkSize * 3;
+
+		if (!$chunkSize || (mb_strlen($prompt) <= $maxChars)) {
+			// Chunking is disabled or prompt is short enough to be a single chunk
+			return [$prompt];
 		}
-		if (count($completion) > 0) {
-			$endTime = time();
-			$this->openAiAPIService->updateExpTextProcessingTime($endTime - $startTime);
-			return ['output' => array_pop($completion)];
+
+		$allChunksValid = function (array $prompts, int $maxChars): bool {
+			// Check that all chunks are no more than maximum chunk size (by number of characters due to use of regular expressions),
+			// number of bytes may still be exceeded and considered separately
+			return array_reduce($prompts, fn (bool $prev, string $next): bool => $prev && (mb_strlen($next) <= $maxChars), true);
+		};
+
+		// Try splitting by paragraph, match as many paragraphs as possible per chunk up to the maximum chunk size
+		$splitByLine = sprintf('/(.{,%d}\n)/su', $maxChars);
+		$prompts = preg_split($splitByLine, $prompt, -1, PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE);
+
+		if ($allChunksValid($prompts, $maxChars)) {
+			return $prompts;
+		}
+
+		// Try splitting by sentence
+		$splitBySentence = sprintf('/(.{,%d}[!\.\?\n])/su', $maxChars);
+		$prompts = preg_split($splitBySentence, $prompt, -1, PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE);
+
+		if ($allChunksValid($prompts, $maxChars)) {
+			return $prompts;
+		}
+
+		// Try splitting by word
+		$splitByWord = sprintf('/(.{,%d}[!\.\?\s])/su', $maxChars);
+		$prompts = preg_split($splitByWord, $prompt, -1, PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE);
+
+		if ($allChunksValid($prompts, $maxChars)) {
+			return $prompts;
 		}
 
-		throw new RuntimeException('No result in OpenAI/LocalAI response.');
+		// Split by number of characters in maximum chunk size
+		$splitBySize = sprintf('/(.{,%d})/su', $maxChars);
+		return preg_split($splitBySize, $prompt, -1, PREG_SPLIT_NO_EMPTY | PREG_SPLIT_DELIM_CAPTURE);
```

Comment:
For some reasons these preg_split calls all return only a single chunk for me having set chunk size to 1000 :thinking:

Reply:
See my comment above:

> Apparently, the discrepancy is due to an outdated version of the PCRE library where having pcre.jit enabled changes the result, and the version bundled in nextcloud-docker-dev by default somehow isn't high enough. I'm not sure if there's an easy way to update the library, but after that update, the splitting should work again.

Reply:
I wonder if we should try using a different method then. I don't know how widespread these circumstances are. I think we can't require people to update pcre.

Reply:
Maybe just using the same regexes but without the capturing groups can work too, I'll need some time to test them again either way.

Reply:
Just pushed a new chunking algorithm which is similar to the previous one and still runs into the same library problem, but is much cleaner when it works. Manually splitting without regexes, especially by sentence, would be a chore and take more time and lines of code to implement. And if the regexes fail, it would just split the string every *n* characters as a fallback.